## Overview
The business has been investing in four different display advertising campaigns. You have been asked to evaluate how effective
each advertising campaign is in generating sales. Please see below the explanation of the data and the business questions.

The data on the left include information on 10000 consumers who clicked on at least one of the display ads from campaigns
A, B, C, or D. Purchase is indicated by the "Conversion" variable (i.e., equals 1 if there is purchase and 0 otherwise). The "Value" column indicates the revenue in dollars earned from each purchase. The cost per click is $7, $5, $4, and $2 for campaign A, B, C, and D, respectively. The order of clicks is as indicated in the data.

You are asked to build a statistical model (so not a rule based attribution model such as last click attribution) and answer:

Questions
1. Which campaign is the most successful in terms of unit sales contributed?

2. What is the return on investment for each campaign?

3. How would you optimize the spend of a given budget of $1 million across all four campaigns?

#### Assume the cost of campaign is only charged when there is a conversion
#### The idea is to construct transition probability matrix between campaigns, to evaluate each campaign effect on conversion units

In [2]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from pulp import *
from collections import defaultdict

from pprint import pprint

import warnings
warnings.filterwarnings('ignore')

In [16]:
df = pd.read_csv("./data/user_click.csv")
columns = ["user", "click_1", "click_2", "click_3", "click_4", "click_5", "conversion", "value"]
df = df[columns]
display(df.head(5))
# Fill missing values with "END" placeholder
df[["click_1", "click_2", "click_3", "click_4", "click_5"]] = df[["click_1", "click_2", "click_3", "click_4", "click_5"]].fillna("END")

# Create separate dataframes for converted and non-converted users
df_converted = df[df['conversion'] == 1]
df_non_converted = df[df['conversion'] == 0]

def calculate_transition_matrix(df):
    transition_counts = defaultdict(lambda: defaultdict(int))
    total_counts = defaultdict(int)

    for _, row in df.iterrows():
        clicks = row[['click_1', 'click_2', 'click_3', 'click_4', 'click_5']].tolist()
        if row['conversion'] == 1:
            clicks.append('conv')
        else:
            clicks.append('END')

        for i in range(len(clicks) - 1):
            current_click = clicks[i]
            next_click = clicks[i + 1]
            transition_counts[current_click][next_click] += 1
            total_counts[current_click] += 1

    transition_matrix = {k: {kk: vv / total_counts[k] for kk, vv in v.items()} for k, v in transition_counts.items()}
    return transition_matrix

# Calculate transition matrices for converted and non-converted users
transition_matrix_converted = calculate_transition_matrix(df_converted)
transition_matrix_non_converted = calculate_transition_matrix(df_non_converted)

# Prepare data
df['path'] = df.apply(lambda x: ['start'] + [x['click_'+str(i)] for i in range(1, 6) 
                                             if str(x['click_'+str(i)]) != 'END'] + ['conv' if x['conversion'] == 1 else 'END']
                                             ,axis=1)

# Calculate transition matrix
transitions = defaultdict(int)
for path in df['path']:
    for i in range(len(path) - 1):
        transitions[(path[i], path[i+1])] += 1

total = sum(transitions.values())
for key in transitions:
    transitions[key] /= total

# Calculate total conversions and transitions from all campaigns
total_conversions = sum([transitions[(campaign, 'conv')] for campaign in ['a', 'b', 'c', 'd']])
total_transitions = sum([sum([transitions[(campaign, next_state)] for next_state in ['a', 'b', 'c', 'd', 'conv', 'END']]) for campaign in ['a', 'b', 'c', 'd']])

# Calculate conversion probability with all campaigns
prob_all = total_conversions / total_transitions

# Calculate removal effects for individual campaigns and their combinations
removal_effects = {}
campaign_combinations = [['a'], ['b'], ['c'], ['d']]
                        #  , ['a', 'b'], ['a', 'c'], ['a', 'd'], ['b', 'c'], ['b', 'd'], ['c', 'd'], ['a', 'b', 'c'], ['a', 'b', 'd'], ['a', 'c', 'd'], ['b', 'c', 'd'], ['a', 'b', 'c', 'd']]

for combination in campaign_combinations:
    transitions_removed = transitions.copy()
    to_remove = [key for key in transitions if any(campaign in key for campaign in combination)]
    
    for key in to_remove:
        del transitions_removed[key]
    
    total_conversions_removed = sum([transitions_removed[(other_campaign, 'conv')] for other_campaign in ['a', 'b', 'c', 'd'] if (other_campaign, 'conv') in transitions_removed])
    total_transitions_removed = sum([sum([transitions_removed[(other_campaign, next_state)] for next_state in ['a', 'b', 'c', 'd', 'conv', 'END'] if (other_campaign, next_state) in transitions_removed]) for other_campaign in ['a', 'b', 'c', 'd']])
    
    prob_removed = total_conversions_removed / total_transitions_removed if total_transitions_removed != 0 else 0
    
    removal_effects[tuple(combination)] = prob_all - prob_removed
print("removal_effects")
pprint(removal_effects)
# Calculate ROI
cpc = {'a': 7, 'b': 5, 'c': 4, 'd': 2}
roi = {}
revenue = df[df.value > 0].value.sum()
total_click_by_campaign = {'a': 0, 'b': 0, 'c': 0, 'd': 0}
for i in range(1, 6):
    click_dict = df[f"click_{i}"].value_counts().to_dict()
    total_click_by_campaign['a'] += click_dict.get('a', 0)
    total_click_by_campaign['b'] += click_dict.get('b', 0)
    total_click_by_campaign['c'] += click_dict.get('c', 0)
    total_click_by_campaign['d'] += click_dict.get('d', 0)

for campaign in ['a', 'b', 'c', 'd']:
    revenue_contribution = revenue * removal_effects[(campaign,)] / sum(removal_effects.values())
    cost = cpc[campaign] * total_click_by_campaign[campaign]
    roi[campaign] = (revenue_contribution - cost) / cost

# Optimization
budget = 1000000  # 1 million dollars
campaigns = ['a', 'b', 'c', 'd']
costs = [cpc[campaign] for campaign in campaigns]
conversions = [removal_effects[(campaign,)] for campaign in campaigns]
revenue_contributions = [revenue * removal_effects[(campaign,)] / sum(removal_effects.values()) for campaign in campaigns]

# Define the optimization problem
# Objective function: Maximize total conversions
# Constraints:
#   - Total cost should be within the budget
#   - Allocations should be non-negative
#   - Revenue contribution of each campaign should be proportional to its removal effect

# Set up the linear programming problem
objective = [-conversion for conversion in conversions]  # Negative sign for maximization

lhs_ineq = [costs]
rhs_ineq = [budget]

lhs_eq = []
rhs_eq = []
for i in range(len(campaigns)):
    lhs_eq.append([0] * len(campaigns))
    lhs_eq[i][i] = revenue_contributions[i]
    rhs_eq.append(revenue_contributions[i])

bnd = [(0, float('inf')) for _ in range(len(campaigns))]  # Non-negative allocations

# Solve the optimization problem
opt_result = linprog(c=objective, A_ub=lhs_ineq, b_ub=rhs_ineq, A_eq=lhs_eq, b_eq=rhs_eq, bounds=bnd, method='simplex')

# Retrieve the optimal allocation
optimized_allocation = {campaigns[i]: round(opt_result.x[i]) for i in range(len(campaigns))}

# Calculate the total conversions, cost, and revenue based on the optimal allocation
total_conversions_optimized = -opt_result.fun
total_cost_optimized = sum([costs[i] * opt_result.x[i] for i in range(len(campaigns))])
total_revenue_optimized = sum([revenue_contributions[i] * opt_result.x[i] for i in range(len(campaigns))])

# Print results
print('Removal effects for individual campaigns:', {k: v for k, v in removal_effects.items() if len(k) == 1})
print('Removal effects for campaign combinations:', {k: v for k, v in removal_effects.items() if len(k) > 1})
print('ROI:', roi)
print('Optimized budget allocation:', optimized_allocation)
print(f'Total conversions with optimized allocation: {total_conversions_optimized}')
print(f'Total cost with optimized allocation: {total_cost_optimized}')
print(f'Total revenue with optimized allocation: {total_revenue_optimized}')

user click_1 click_2 click_3 click_4 click_5  conversion  value
0     1       d       d       a       b     NaN           0      0
1     2       c     NaN     NaN     NaN     NaN           0      0
2     3       b       d       a       b       c           1     44
3     4       b       a       a     NaN     NaN           0      0
4     5       d       b       d     NaN     NaN           0      0

removal_effects
{('a',): -0.010831787213580885,
 ('b',): -0.015342039228628826,
 ('c',): -0.014694348312304817,
 ('d',): -0.017900802875224858}
Removal effects for individual campaigns: {('a',): -0.010831787213580885, ('b',): -0.015342039228628826, ('c',): -0.014694348312304817, ('d',): -0.017900802875224858}
Removal effects for campaign combinations: {}
ROI: {'a': -0.6855895345577628, 'b': -0.396536129266172, 'c': -0.2601074168630421, 'd': 0.8144123888477783}
Optimized budget allocation: {'a': 1, 'b': 1, 'c': 1, 'd': 1}
Total conversions with optimized allocation: -0.058768977629739386
Total cost with optimized allocation: 18.0
Total revenue with optimized allocation: 87934.0


In [7]:
removal_effects[(campaign,)] / sum(removal_effects.values())

0.3045961253232072

In [14]:
opt_result

 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: 0.0
       x: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00]
     nit: 2

In [11]:
conversions

[-0.010831787213580885,
 -0.015342039228628826,
 -0.014694348312304817,
 -0.017900802875224858]

0.1843113093071672
0.26105676578700865
0.25003579958261696
0.30459612532320723
Removal effects: {'a': -0.010831787213580885, 'b': -0.015342039228628826, 'c': -0.014694348312304817, 'd': -0.017900802875224858}
ROI: {'a': -0.6855895345577628, 'b': -0.396536129266172, 'c': -0.2601074168630421, 'd': 0.8144123888477783}


In [18]:
import pandas as pd
import numpy as np
from collections import defaultdict
from scipy.sparse import csr_matrix
from scipy.sparse.csgraph import connected_components
from scipy.sparse.csgraph import dijkstra

# Load data
# data = pd.read_csv('data.csv')

# Prepare data
df['path'] = df.apply(lambda x: ['start'] + [x['click_'+str(i)] if str(x['click_'+str(i)]) != 'NULL' else 'END' for i in range(1, 6)] + ['conv' if x['conversion'] == 1 else 'NULL'], axis=1)

# Calculate transition matrix
transitions = defaultdict(int)
for path in df['path']:
    for i in range(len(path) - 1):
        transitions[(path[i], path[i+1])] += 1

total = sum(transitions.values())
for key in transitions:
    transitions[key] /= total

# Create a list of all states
states = list(set([state for (state_from, state_to) in transitions.keys()] + ['conv', 'END']))

# Create a transition matrix
transition_matrix = pd.DataFrame(data=np.zeros((len(states), len(states))), index=states, columns=states)
for (state_from, state_to), prob in transitions.items():
    transition_matrix.at[state_from, state_to] = prob

# Calculate removal effects
removal_effects = {}
for campaign in ['a', 'b', 'c', 'd']:
    # Create a transition matrix without campaign
    transition_matrix_removed = transition_matrix.drop(index=campaign, columns=campaign).copy()
    
    # Redistribute transitions of removed campaign
    for state in ['a', 'b', 'c', 'd', 'conv', 'END']:
        if state != campaign:
            transition_matrix_removed.loc[state] += transition_matrix.loc[campaign] * transition_matrix.loc[state].sum()
            transition_matrix_removed[state] += transition_matrix[campaign] * transition_matrix[state].sum()
    
    # Normalize transition matrix
    transition_matrix_removed = transition_matrix_removed.div(transition_matrix_removed.sum(axis=1), axis=0)
    
    # Calculate removal effect
    removal_effects[campaign] = transition_matrix.at['start', 'conv'] - transition_matrix_removed.at['start', 'conv']

# Print results
print('Removal effects:', removal_effects)


NameError: name 'state' is not defined

In [17]:
transitions.keys()

dict_keys([('start', 'd'), ('d', 'd'), ('d', 'a'), ('a', 'b'), ('b', 'END'), ('END', 'NULL'), ('start', 'c'), ('c', 'END'), ('END', 'END'), ('start', 'b'), ('b', 'd'), ('b', 'c'), ('c', 'conv'), ('b', 'a'), ('a', 'a'), ('a', 'END'), ('d', 'b'), ('d', 'END'), ('c', 'b'), ('a', 'c'), ('start', 'a'), ('d', 'c'), ('c', 'a'), ('a', 'd'), ('c', 'NULL'), ('END', 'conv'), ('c', 'd'), ('b', 'b'), ('c', 'c'), ('b', 'NULL'), ('d', 'NULL'), ('d', 'conv'), ('a', 'NULL'), ('a', 'conv'), ('b', 'conv')])

In [39]:
# Cost per click for each campaign
cost_per_click = {'a': 7, 'b': 5, 'c': 4, 'd': 2}

# Calculate total clicks and revenue for each campaign
total_clicks = {campaign: (df == campaign).sum().sum() for campaign in ['a', 'b', 'c', 'd']}
total_revenue = {campaign: df[df[['click_1', 'click_2', 'click_3', 'click_4', 'click_5']].apply(lambda row: campaign in row.values, axis=1)]['value'].sum() for campaign in ['a', 'b', 'c', 'd']}

# Calculate ROI for each campaign
roi = {campaign: (total_revenue[campaign] - total_clicks[campaign] * cost_per_click[campaign]) / (total_clicks[campaign] * cost_per_click[campaign]) for campaign in ['a', 'b', 'c', 'd']}
pprint(roi)


{'a': 0.12970435322417942,
 'b': 0.44348054679284965,
 'c': 0.818481626060035,
 'd': 2.384433003658041}


In [40]:
# Total budget
total_budget = 1_000_000

# Proportion of budget based on ROI
total_roi = sum(roi.values())
budget_allocation = {campaign: (roi[campaign] / total_roi) * total_budget for campaign in roi}

# Print budget allocation
print("Budget allocation based on ROI:")
for campaign, budget in budget_allocation.items():
    print(f"Campaign {campaign}: ${budget:.2f}")


Budget allocation based on ROI:
Campaign a: $34348.76
Campaign b: $117444.08
Campaign c: $216753.19
Campaign d: $631453.96


In [41]:
roi

{'a': 0.12970435322417942,
 'b': 0.44348054679284965,
 'c': 0.818481626060035,
 'd': 2.384433003658041}

In [42]:
df[df.conversion > 0].shape

(2203, 8)

#### DEFINE STATES
##### START, a, b, c, d, NULL, Conversion

In [4]:
#The cost per click is $7, $5, $4, and $2 for campaign A, B, C, and D, respectively. 
# The company pay for the cost per click
# the profit = revenue 
cpc_dict = {"a": 7, "b": 5, "c": 4, "d": 2}
df["cost"] = 0
for i in range(1, 6):
    df[f"click_{i}_cost"] = df[f"click_{i}"].map(cpc_dict).fillna(0)
    df["cost"] += df[f"click_{i}_cost"]
df["profit"] = df["value"] - df["cost"]
df.profit[df.profit < 0] = 0
df["path"] = df[["click_1", "click_2", "click_3", "click_4", "click_5"]].astype(str).agg(" ".join, axis=1) #.str.split(" ")

In [5]:
df['path'] = np.where(df['conversion'] == 0,
                      ['Start '] + df['path'] + [' NULL'],
                      ['Start '] + df['path'] + [' Conversion'])
df["path"] = df["path"].str.split(" ")

In [6]:
df.head(5)

user click_1 click_2 click_3 click_4 click_5  conversion  value  cost  \
0     1       d       d       a       b    NULL           0      0  16.0   
1     2       c    NULL    NULL    NULL    NULL           0      0   4.0   
2     3       b       d       a       b       c           1     44  23.0   
3     4       b       a       a    NULL    NULL           0      0  19.0   
4     5       d       b       d    NULL    NULL           0      0   9.0   

   click_1_cost  click_2_cost  click_3_cost  click_4_cost  click_5_cost  \
0             2           2.0           7.0           5.0           0.0   
1             4           0.0           0.0           0.0           0.0   
2             5           2.0           7.0           5.0           4.0   
3             5           7.0           7.0           0.0           0.0   
4             2           5.0           2.0           0.0           0.0   

   profit                                      path  
0     0.0           [Start, d, d, a, b, NULL, NULL]  
1     0.0  [Start, c, NULL, NULL, NULL, NULL, NULL]  
2    21.0        [Start, b, d, a, b, c, Conversion]  
3     0.0        [Start, b, a, a, NULL, NULL, NULL]  
4     0.0        [Start, d, b, d, NULL, NULL, NULL]

In [7]:
df.cost.sum(), df.value.sum(), df.profit.sum()

(134066.0, 87934, 50954.0)

In [8]:
list_of_paths = df['path']
total_conversions = sum(path.count('Conversion') for path in df['path'].tolist())
base_conversion_rate = total_conversions / len(list_of_paths)
total_value = df["value"].sum()
base_revenue_rate = total_value / len(list_of_paths)
total_profit = df["profit"].sum()
base_profit_rate = total_profit / len(list_of_paths)

#### BUILD TRANSITION STATES

In [9]:
def transition_states(list_of_paths):
    list_of_unique_channels = set(x for element in list_of_paths for x in element)
    #print(list_of_unique_channels)
    transition_states = {x + '-->' + y: 0 for x in list_of_unique_channels for y in list_of_unique_channels}
    #print(transition_states)
    for possible_state in list_of_unique_channels:
        if possible_state not in ['Conversion', 'NULL']:
            for user_path in list_of_paths:
                if possible_state in user_path:
                    indices = [i for i, s in enumerate(user_path) if possible_state in s]
                    for col in indices:
                        #print(col)
                        #print(user_path)
                        #print(user_path[col])
                        if col < len(user_path) - 1:
                            transition_states[user_path[col] + '-->' + user_path[col + 1]] += 1

    return transition_states

In [10]:
trans_states = transition_states(list_of_paths)

In [11]:
trans_states

{'Conversion-->Conversion': 0,
 'Conversion-->a': 0,
 'Conversion-->d': 0,
 'Conversion-->c': 0,
 'Conversion-->NULL': 0,
 'Conversion-->b': 0,
 'Conversion-->Start': 0,
 'a-->Conversion': 183,
 'a-->a': 1226,
 'a-->d': 1214,
 'a-->c': 1212,
 'a-->NULL': 2243,
 'a-->b': 1286,
 'a-->Start': 0,
 'd-->Conversion': 157,
 'd-->a': 1219,
 'd-->d': 1223,
 'd-->c': 1212,
 'd-->NULL': 2334,
 'd-->b': 1236,
 'd-->Start': 0,
 'c-->Conversion': 165,
 'c-->a': 1168,
 'c-->d': 1228,
 'c-->c': 1256,
 'c-->NULL': 2356,
 'c-->b': 1256,
 'c-->Start': 0,
 'NULL-->Conversion': 0,
 'NULL-->a': 0,
 'NULL-->d': 0,
 'NULL-->c': 0,
 'NULL-->NULL': 0,
 'NULL-->b': 0,
 'NULL-->Start': 0,
 'b-->Conversion': 192,
 'b-->a': 1258,
 'b-->d': 1259,
 'b-->c': 1252,
 'b-->NULL': 2370,
 'b-->b': 1277,
 'b-->Start': 0,
 'Start-->Conversion': 0,
 'Start-->a': 4986,
 'Start-->d': 3403,
 'Start-->c': 3426,
 'Start-->NULL': 0,
 'Start-->b': 3513,
 'Start-->Start': 0}

#### FROM TRANSITION FREQUENCY, BUILD TRANSITION PROBABILITY MATRIX

In [12]:
def transition_probability(trans_dict):
    list_of_unique_channels = set(x for element in list_of_paths for x in element)
    trans_prob = defaultdict(dict)
    for state in list_of_unique_channels:
        if state not in ['Conversion', 'NULL']:
            counter = 0
            index = [i for i, s in enumerate(trans_dict) if state + '-->' in s]
            for col in index:
                if trans_dict[list(trans_dict)[col]] > 0:
                    counter += trans_dict[list(trans_dict)[col]]
            for col in index:
                if trans_dict[list(trans_dict)[col]] > 0:
                    state_prob = float((trans_dict[list(trans_dict)[col]])) / float(counter)
                    trans_prob[list(trans_dict)[col]] = state_prob

    return trans_prob

In [13]:
state_probability = transition_probability(trans_states)
pprint(state_probability)

defaultdict(<class 'dict'>,
            {'Start-->a': 0.32528705636743216,
             'Start-->b': 0.2291884133611691,
             'Start-->c': 0.2235125260960334,
             'Start-->d': 0.22201200417536535,
             'a-->Conversion': 0.024850624660510592,
             'a-->NULL': 0.30458989679521997,
             'a-->a': 0.16648560564910375,
             'a-->b': 0.1746333514394351,
             'a-->c': 0.1645844649646931,
             'a-->d': 0.1648560564910375,
             'b-->Conversion': 0.025236593059936908,
             'b-->NULL': 0.3115141955835962,
             'b-->a': 0.16535226077812828,
             'b-->b': 0.16784963196635122,
             'b-->c': 0.16456361724500526,
             'b-->d': 0.16548370136698212,
             'c-->Conversion': 0.02221025710055189,
             'c-->NULL': 0.3171355498721228,
             'c-->a': 0.15722169874814915,
             'c-->b': 0.1690671692017768,
             'c-->c': 0.1690671692017768,
             'c-->d': 0.

In [14]:
def construct_transition_matrix(list_of_paths, state_probability):
    transition_matrix = pd.DataFrame()
    list_of_unique_channels = set(x for element in list_of_paths for x in element)

    for channel in list_of_unique_channels:
        transition_matrix[channel] = 0.00
        transition_matrix.loc[channel] = 0.00
        transition_matrix.loc[channel][channel] = 1.0 if channel in ['Conversion', 'NULL'] else 0.0

    for key, value in state_probability.items():
        origin, destination = key.split('-->')
        transition_matrix.at[origin, destination] = value

    return transition_matrix

transition_matrix = construct_transition_matrix(list_of_paths, state_probability)
print("Markov transition matrix")
display(transition_matrix)

Markov transition matrix


Conversion         a         d         c      NULL         b  \
Conversion    1.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
a             0.024851  0.166486  0.164856  0.164584  0.304590  0.174633   
d             0.021271  0.165154  0.165696  0.164205  0.316217  0.167457   
c             0.022210  0.157222  0.165298  0.169067  0.317136  0.169067   
NULL          0.000000  0.000000  0.000000  0.000000  1.000000  0.000000   
b             0.025237  0.165352  0.165484  0.164564  0.311514  0.167850   
Start         0.000000  0.325287  0.222012  0.223513  0.000000  0.229188   

            Start  
Conversion    0.0  
a             0.0  
d             0.0  
c             0.0  
NULL          0.0  
b             0.0  
Start         0.0

#### HAVING TRANSITION MATRIX, CALCULATE REMOVAL EFFECTS OF EACH CAMPAIGN

In [15]:
def removal_effects(df, conversion_rate):
    removal_effects_dict = {}
    channels = [channel for channel in df.columns if channel not in ['Start',
                                                                     'NULL',
                                                                     'Conversion']]
    for channel in channels:
        removal_df = df.drop(channel, axis=1).drop(channel, axis=0)
        for column in removal_df.columns:
            row_sum = np.sum(list(removal_df.loc[column]))
            null_pct = float(1) - row_sum
            if null_pct != 0:
                removal_df.loc[column]['NULL'] = null_pct
            removal_df.loc['NULL']['NULL'] = 1.0

        removal_to_conv = removal_df[
            ['NULL', 'Conversion']].drop(['NULL', 'Conversion'], axis=0)
        removal_to_non_conv = removal_df.drop(
            ['NULL', 'Conversion'], axis=1).drop(['NULL', 'Conversion'], axis=0)

        removal_inv_diff = np.linalg.inv(
            np.identity(
                len(removal_to_non_conv.columns)) - np.asarray(removal_to_non_conv))
        removal_dot_prod = np.dot(removal_inv_diff, np.asarray(removal_to_conv))
        removal_cvr = pd.DataFrame(removal_dot_prod,
                                   index=removal_to_conv.index)[[1]].loc['Start'].values[0]
        removal_effect = 1 - removal_cvr / conversion_rate
        removal_effects_dict[channel] = removal_effect

    return {k: v for k, v in sorted(removal_effects_dict.items())} 

campaign_removal_effect = removal_effects(transition_matrix, base_conversion_rate)

In [16]:
def allocator(removal_effects, unit):
    sum_effect = sum(list(removal_effects.values()))    
    allocation_result = {k: (v / sum_effect) * unit for k, v in sorted(removal_effects.items())}    
    return allocation_result

conversion_unit_attributions = allocator(campaign_removal_effect, 1)

#### RESULT OF CONVERSTION ATTRIBUTION

In [17]:
removal_effects_dict = removal_effects(transition_matrix, base_conversion_rate)
revenue_attributions = allocator(removal_effects_dict, total_value)
print("removal_effects_dict")
print(removal_effects_dict)
print("Conversion Unit Attribution")
print(conversion_unit_attributions)
print("revenue attributions")
print(revenue_attributions)

removal_effects_dict
{'a': 0.8597100706463602, 'b': 0.8414695477652324, 'c': 0.8312865403541212, 'd': 0.8293959429065695}
Conversion Unit Attribution
{'a': 0.2557243707940063, 'b': 0.25029865066347073, 'c': 0.24726967234635122, 'd': 0.24670730619617173}
revenue attributions
{'a': 22486.86682140015, 'b': 22009.761547441634, 'c': 21743.411368104047, 'd': 21693.960263054167}


#### Cost per campaign
(again, assume the cost is charged when there is conversion)

In [18]:
cost_campaign = {}
for campaign in ["a", "b", "c", "d"]:
    cost_campaign[campaign] = 0
    for n_click in range(1, 6):
        cost_campaign[campaign] += df.loc[((df[f"click_{n_click}"]==campaign)
                                           &(df["value"] > 0)), f"click_{n_click}_cost"].sum()

In [19]:
cost_campaign

{'a': 15470.0, 'b': 10060.0, 'c': 7916.0, 'd': 3542.0}

### ROI per campaign

In [20]:
roi_attributions = {}
for k, v in revenue_attributions.items():
    roi_attributions[k] = v - cost_campaign[k]

In [21]:
roi_attributions

{'a': 7016.866821400148,
 'b': 11949.761547441634,
 'c': 13827.411368104047,
 'd': 18151.960263054167}

In [27]:
for k, v in revenue_attributions.items():
    print(k, round(100 * roi_attributions[k] / revenue_attributions[k], 2), "%")

a 31.2 %
b 54.29 %
c 63.59 %
d 83.67 %


#### Budget Allocation

In [22]:
budget = 1e6
budget_attributions = allocator(removal_effects_dict, budget)

In [23]:
print(budget_attributions)

{'a': 255724.37079400627, 'b': 250298.65066347073, 'c': 247269.67234635123, 'd': 246707.30619617173}


### CONCLUSIONS
- Campaign a is the highest contributor of unit sales
- RoI of each campaign:
   - a ~ 7016 (31.2%)
   - b ~ 11949 (54.3%)
   - c ~ 13827 (63.6%)
   - d ~ 18151 (83.7%)
- For budget allocation, the distribution: a / b / c / d = 25.6 / 25 / 24.7 / 24.7 (in %)